src/lib/protocols/ipsec.c

```diff
@@ -0,0 +1,198 @@
+  {
+    return ndpi_int_check_ports(packet);
+  }
+
+  /* Flags */
+  if (packet->payload[isakmp_offset + 19] >= 8)
+  {
+    return ndpi_int_check_ports(packet);
+  }
+
+  return isakmp_type;
+}
+
+static enum isakmp_type ndpi_int_check_isakmp_v2(struct ndpi_packet_struct const * const packet,
+                                                 u_int16_t isakmp_offset, enum isakmp_type isakmp_type)
+{
+  /* Next payload type */
+  if ((packet->payload[isakmp_offset + 16] > 0 && packet->payload[isakmp_offset + 16] <= 32) ||
+      (packet->payload[isakmp_offset + 16] >= 49 && packet->payload[isakmp_offset + 16] <= 127))
+  {
+    return ndpi_int_check_ports(packet);
+  }
+
+  /* Exchange Type */
+  if ((packet->payload[isakmp_offset + 18] <= 33) ||
+      (packet->payload[isakmp_offset + 18] >= 38 && packet->payload[isakmp_offset + 18] <= 239))
+  {
+    return ndpi_int_check_ports(packet);
+  }
+
+  /* Flags */
+  if ((packet->payload[isakmp_offset + 19] & 0xC7) != 0)
+  {
+    return ndpi_int_check_ports(packet);
+  }
+
+  return isakmp_type;
+}
+
+void ndpi_search_ipsec(struct ndpi_detection_module_struct *ndpi_struct,
+                       struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct * const packet = &ndpi_struct->packet;
+  u_int16_t isakmp_offset = 0;
+  enum isakmp_type isakmp_type = ISAKMP_INVALID;
+
+  NDPI_LOG_DBG(ndpi_struct, "search IPSEC (UDP)\n");
+
+  if (packet->payload_packet_len < 28)
+  {
+    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+    return;
+  }
+
+  /* check for non-ESP marker required for ISAKMP over UDP */
+  if (get_u_int32_t(packet->payload, 0) == 0x00000000)
+  {
+    isakmp_offset = 4;
+    if (packet->payload_packet_len < 32)
+    {
+      NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+      return;
+    }
+  }
+
+  if (packet->payload[isakmp_offset + 17] != 0x20 /* Major Version 2 */)
+  {
+    if (packet->payload[isakmp_offset + 17] != 0x10 /* Major Version 1 */)
+    {
+      NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+      return;
+    } else {
+      /* Version 1 is obsolete, but still used by some embedded devices. */
+      isakmp_type = ISAKMP_V1;
+    }
+  } else {
+    isakmp_type = ISAKMP_V2;
+  }
+
+  if (ntohl(get_u_int32_t(packet->payload, isakmp_offset + 24)) != (u_int32_t)packet->payload_packet_len - isakmp_offset)
+  {
+    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+    return;
+  }
+
+  if (isakmp_type == ISAKMP_V1)
+  {
+    isakmp_type = ndpi_int_check_isakmp_v1(packet, isakmp_offset, isakmp_type);
+  } else {
+    isakmp_type = ndpi_int_check_isakmp_v2(packet, isakmp_offset, isakmp_type);
+  }
+
+  ndpi_int_ipsec_add_connection(ndpi_struct, flow, isakmp_type);
+}
+
+void init_ipsec_dissector(struct ndpi_detection_module_struct *ndpi_struct,
+                          u_int32_t *id, NDPI_PROTOCOL_BITMASK *detection_bitmask)
+{
+  ndpi_set_bitmask_protocol_detection("IPSEC", ndpi_struct, detection_bitmask, *id,
```

Comment:
... "IPSEC". Should we use the same name?

Reply:
definitly, let's stick to ÌPSec`